src/main/java/cceclipseplugin/editor/listeners/DocumentChangeListener.java

```diff
@@ -82,12 +83,17 @@ public void documentAboutToBeChanged(DocumentEvent event) {
 
 		List<Diff> newDiffs = new ArrayList<>();
 		diffLoop: for (int i = 0; i < diffs.size(); i++) {
-			while (!docMgr.getAppliedDiffs().isEmpty()) {
-				Diff appliedDiff = docMgr.getAppliedDiffs().poll();
-				System.out.printf("DEBUG SEND-ON-NOTIF: %s ?= %s; %b\n", diffs.get(i).toString(),
-						appliedDiff.toString(), diffs.get(i).equals(appliedDiff));
-				if (diffs.get(i).equals(appliedDiff)) {
-					continue diffLoop;
+			synchronized(docMgr.getAppliedDiffs(currFile)){
+				for(int j = 0; j < docMgr.getAppliedDiffs(currFile).size(); j++){
+					Diff appliedDiff = docMgr.getAppliedDiffs(currFile).get(j);
+					System.out.printf("DEBUG SEND-ON-NOTIF: %s ?= %s; %b\n", diffs.get(i).toString(),
+							appliedDiff.toString(), diffs.get(i).equals(appliedDiff));
+					if(appliedDiff.equals(diffs.get(i))){
+						for(int k = j-1; k >= 0; k--){
+							docMgr.getAppliedDiffs(currFile).removeFirst();
+						}
+						continue diffLoop;
```

Comment:
`docMgr.getAppliedDiffs` is called 4 times in the last 10 lines. Capture the applied diffs in a local variable right before the synchronized block.